src/box/lua/connpool.lua

```diff
@@ -271,46 +316,102 @@ local function filter(opts)
         if is_candidate_match_dynamic(instance_name, dynamic_opts) then
             table.insert(dynamic_candidates, instance_name)
         end
     end
+
     return dynamic_candidates
 end
 
+-- Subroutine for the `call()` function. Tries to find an active
+-- connection matching the requirements and considering the
+-- specified priorities.
 local function get_connection(opts)
-    local mode = nil
+    assert(opts.mode == 'alive' or opts.mode == 'ro' or opts.mode == 'rw' or
+           opts.mode == 'prefer_ro' or opts.mode == 'prefer_rw')
+
+    local mode = 'alive'
     if opts.mode == 'ro' or opts.mode == 'rw' then
         mode = opts.mode
     end
-    local candidates_opts = {
+
+    -- It's better to use the local instance to perform the call
+    -- faster. So prefer the local instance by default.
+    local prefer_local = true
+    if opts.prefer_local ~= nil then
+        prefer_local = opts.prefer_local
+    end
+
+    local dynamic_opts = {
+        mode = mode,
+    }
+
+    local local_instance_name = box.info.name
+
+    local filter_static_opts = {
         groups = opts.groups,
         replicasets = opts.replicasets,
         instances = opts.instances,
         labels = opts.labels,
         roles = opts.roles,
         sharding_roles = opts.sharding_roles,
-        mode = mode,
     }
-    local candidates = filter(candidates_opts)
+    local static_candidates = filter(filter_static_opts)
+
+    -- In case we don't need to prioritize the matching instances
+    -- we can wait for any connection available.
+    if opts.mode ~= 'prefer_rw' and opts.mode ~= 'prefer_ro' then
+        if prefer_local then
+            for _, instance_name in ipairs(static_candidates) do
+                if instance_name ~= local_instance_name then
+                    goto continue
+                end
+
+                local conn = connect(local_instance_name,
+                                     { wait_connected = true })
+                assert(conn.state == 'active')
+
+                if is_candidate_match_dynamic(local_instance_name,
+                                              dynamic_opts) then
+                    return conn
+                end
+                ::continue::
+            end
+        end
+
+        local candidate = acquire_any_candidate(static_candidates,
+                                                dynamic_opts)
+
+        if not candidate then
+            return nil, "no candidates are available with these conditions"
+        end
+
+        return candidate
+    end
+
+    -- In case there are specified prioritizing of matching
+    -- instances we fetch all matched static candidates.
+    -- Technically, it's possible to speedup fetching the instances
+    -- if there is already a connection to a top-priority connection
+    -- (e.g. one RO instance is established and the mode is
+    -- 'prefer_ro')
+    local candidates = acquire_all_candidates(static_candidates)
     if next(candidates) == nil then
```

Comment:
It seems meaningful to exit earlier if there are no static candidates. Now the check is done only after connecting. The reason is that `acquire_all_candidates` is no-op on an empty list of candidates?

Reply:
Since I've made `acquire_all_candidates()` return empty list in case there is no parameters it works ok now.

We get to that line in case of `mode = prefer_*` and if there is no candidates the `acquire_all_candidates()` instantly returns.